### PR TITLE
[LLVM-Reduce] - Distinct Metadata Reduction

### PR DESCRIPTION
**Pull request by rbintel**
- Add a distinct metadata reduction pass, which traverses the whole unnamed metadata tree and applies reduction where possible. Previous version could do this only partially, either removing named metadata, metadata attached to instructions or debug information.

- Add the --aggressive-named-md-reduction flag, which allows the reducer to reduce named metadata operands more aggressively.

- Add tests for the new functionality.

### CODE REVIEW

**Comment by rbintel** [comment]
@arsenm This is the replacement PR for [96072](https://github.com/llvm/llvm-project/pull/96072), addressed the issues which you pointed out in the review.

File: llvm/test/tools/llvm-reduce/Inputs/remove-metadata.py
Comment (arsenm) [inline llvm/test/tools/llvm-reduce/Inputs/remove-metadata.py:1]: Empty file left behind?

File: llvm/tools/llvm-reduce/deltas/ReduceDistinctMetadata.cpp [hunk 1]
```diff
@@ -0,0 +1,147 @@
...
+#include "llvm/ADT/SetVector.h"
+#include "llvm/ADT/SmallVector.h"
+#include "llvm/IR/InstIterator.h"
+#include <algorithm>
+#include <queue>
+
+using namespace llvm;
+
+// Traverse the graph breadth-first and try to remove unnamed metadata nodes
+void reduceNodes(MDNode *Root,
```
Comment (arsenm) [inline llvm/tools/llvm-reduce/deltas/ReduceDistinctMetadata.cpp:26]: static

File: llvm/tools/llvm-reduce/deltas/ReduceDistinctMetadata.cpp [hunk 2]
```diff
@@ -0,0 +1,147 @@
...
+      if (!O.shouldKeep())
+        Node->replaceOperandWith(PositionToReplace, TemporaryNode);
+    }
+    NodesToDelete.clear();
+  }
+}
+
+// After reducing metadata, we need to remove references to the temporary node,
+// this is also done with BFS
+void cleanUpTemporaries(NamedMDNode &NamedNode, MDTuple *TemporaryTuple,
```
Comment (arsenm) [inline llvm/tools/llvm-reduce/deltas/ReduceDistinctMetadata.cpp:65]: static

File: llvm/tools/llvm-reduce/deltas/ReduceMetadata.cpp [hunk 3]
```diff
@@ -44,24 +51,32 @@ static constexpr StringLiteral ListNamedMetadata[] = {
...
+    // If we don't want to reduce mindlessly, check if our node is part of
+    // ListNamedMetadata before reducing it
+    if (!AggressiveMetadataReduction) {
+      bool Found = false;
+      for (StringRef MDName : ListNamedMetadata) {
+        if (I.getName() == MDName)
+          Found = true;
+      }
+      if (!Found)
+        continue;
```
Comment (arsenm) [inline llvm/tools/llvm-reduce/deltas/ReduceMetadata.cpp:64]: Use is_contained?

File: llvm/tools/llvm-reduce/deltas/ReduceDistinctMetadata.cpp [hunk 4]
```diff
@@ -0,0 +1,147 @@
...
+  while (!NodesToTraverse.empty()) {
+    MDTuple *CurrentTuple = NodesToTraverse.front();
+    NodesToTraverse.pop();
+
+    // Shift all of the interesting elements to the left, pop remaining
+    // afterwards
+    if (CurrentTuple
+            ->isDistinct()) { // Do resizing and cleaning operations only if
+                              // the node is distinct, as resizing is not
+                              // supported for unique nodes and is redundant.
```
Comment (arsenm) [inline llvm/tools/llvm-reduce/deltas/ReduceDistinctMetadata.cpp:91]: Move the comment inside the condition? The wrapping is weird

File: llvm/tools/llvm-reduce/deltas/ReduceDistinctMetadata.cpp [hunk 5]
```diff
@@ -0,0 +1,147 @@
...
+    for (unsigned int I = 0; I < CurrentNode->getNumOperands(); ++I) {
+      if (MDNode *Operand =
+              dyn_cast<MDNode>(CurrentNode->getOperand(I).get())) {
+        // Check whether node has been visited
+        if (!VisitedNodes.contains(Operand)) {
+          NodesToTraverse.push(Operand);
+          VisitedNodes.insert(Operand);
+        }
+        // Delete the node only if it is distinct
+        if (Operand->isDistinct())
```
Comment (arsenm) [inline llvm/tools/llvm-reduce/deltas/ReduceDistinctMetadata.cpp:48]: Braces ,or remove the "Add to removal list" comment

File: llvm/tools/llvm-reduce/deltas/ReduceDistinctMetadata.cpp [hunk 6]
```diff
@@ -0,0 +1,147 @@
...
+    }
+  }
+}
+
+static void extractDistinctMetadataFromModule(Oracle &O,
+                                              ReducerWorkItem &WorkItem) {
+  Module &Program = WorkItem.getModule();
+  MDTuple *TemporaryTuple = MDTuple::getDistinct(
+      Program.getContext(), SmallVector<Metadata *, 1>{llvm::MDString::get(
+                                Program.getContext(), "temporary_tuple")});
```
Comment (arsenm) [inline llvm/tools/llvm-reduce/deltas/ReduceDistinctMetadata.cpp:130]: can you avoid this temporary name? What if something already exists named with "temporary_tuple"?

File: llvm/test/tools/llvm-reduce/remove-metadata.ll [hunk 7]
```diff
@@ -1,8 +1,8 @@
 ; Test that llvm-reduce can remove uninteresting metadata from an IR file.
 ; The Metadata pass erases named & unnamed metadata nodes.
 ;
-; RUN: llvm-reduce --test %python --test-arg %p/Inputs/remove-metadata.py %s -o %t
-; RUN: cat %t | FileCheck -implicit-check-not=! %s
+; RUN: llvm-reduce --aggressive-md --test %python --test-arg %p/Inputs/remove-metadata.py %s -o %t
+; RUN: FileCheck -implicit-check-not=! %s < %t
```
Comment (arsenm) [inline llvm/test/tools/llvm-reduce/remove-metadata.ll:5]: Check with and without the aggressive flag

File: llvm/tools/llvm-reduce/deltas/ReduceMetadata.cpp [hunk 8]
```diff
@@ -20,6 +20,13 @@
 
 using namespace llvm;
 
+extern cl::OptionCategory LLVMReduceOptions;
+
+static cl::opt<bool> AggressiveMetadataReduction(
+    "aggressive-md",
```
Comment (arsenm) [inline llvm/tools/llvm-reduce/deltas/ReduceMetadata.cpp:26]: Make name more specific? Maybe reduce-unknown-named-metadata?

File: llvm/tools/llvm-reduce/deltas/ReduceDistinctMetadata.cpp (hunk 4 above)
Comment (rbintel) [reply llvm/tools/llvm-reduce/deltas/ReduceDistinctMetadata.cpp]: Found the wrapping weird too. It comes from clang-format which is recommended to use by LLVM GitHub user guide

**Comment by rbintel** [comment]
Hey, @arsenm! Addressed all the issues pointed out by the review, should be okay now?

**Review by arsenm — APPROVED** [review]

File: llvm/tools/llvm-reduce/deltas/ReduceDistinctMetadata.cpp [hunk 9]
```diff
@@ -0,0 +1,143 @@
+//===- ReduceDistinctMetadata.cpp - Specialized Delta Pass ------------------------===//
```
Comment (arsenm) [inline llvm/tools/llvm-reduce/deltas/ReduceDistinctMetadata.cpp:1]: clang-format is wrongly breaking these lines, but they are too long

**Review by arsenm — APPROVED** [review]

**Comment by rbintel** [comment]
@arsenm Would it be possible to merge now as all checks were successful?

**Pull request merged**